### ISSUE TITLE
Improve linking error diagnostics

### ISSUE THREAD

**Issue by densh** [scala-native/scala-native#1563]
Currently, whenever linking step fails, it output a sequence of missing symbols. This can be rather unhelpful because it doesn't tell you what piece of code depended on missing symbols.

Instead, for each missing symbol we should output a stacktrace-like summary of why was this symbol required. So for example if A depends on B depends on C, instead of saying we couldn't find C, we would output something along the lines of:

```
* missing symbol C
   - referenced from B
   - referenced from A
```

**Comment by densh** [comment]
This should be really easy to add to 0.4.x but it needs further investigation for 0.3.x (implementation will be different due to rewritten linking and reachability implementation).

**Comment by sjrd** [comment]
See [`Analysis.logError` in Scala.js](https://github.com/scala-js/scala-js/blob/c38ccaf9c561ba6a7d381c7d1c2ee78170812cac/linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analysis.scala#L218) for prior art.

**Comment by densh** [comment]
Fixed for both 0.3.x and 0.4.x now.

**Comment by ekrich** [comment]
Faster gun in the old west!
